Use boost::iostreams::mapped_file_source to improve CSV loading performance [ANT-3915]

a-zakir opened the pull request.

a-zakir committed:
load perf

a-zakir committed:
custom MappedFile 1/2

a-zakir committed:
boost memory Mapped File 2/2

a-zakir committed:
upd

a-zakir committed:
use ub22's default gcc

a-zakir committed:
fix sonarcloud.yml

a-zakir committed:
move time Series Set

a-zakir committed:
upd

a-zakir committed:
upd

a-zakir committed:
upd

a-zakir committed:
upd

guilpier-code commented on src/io/inputs/data-series-csv-importer/DataSeriesRepoImporter.cpp:
```diff
@@ -155,7 +181,8 @@ DataSeriesRepository DataSeriesRepoImporter::importFromDirectory(const std::file
         {
             continue;
         }
-        auto timeSeriesSet = std::make_unique<TimeSeriesSet>(importFromFile(entry, csvSeparators));
+        auto timeSeriesSet = std::make_unique<TimeSeriesSet>(entry.path().stem().string(),
+                                                             readCSV(entry, csvSeparator));
         repo.addDataSeries(std::move(timeSeriesSet));
     }
```
We could even have : 
```c++
for (auto paths = std::filesystem::directory_iterator{path};
     const auto& entry: paths | pathFilter | hasRightExtension)
{
    auto timeSeriesSet = std::make_unique<TimeSeriesSet>(entry.path().stem().string(),
                                                             readCSV(entry, csvSeparator));
    ...
}
```
or something like that

flomnes replied:
We had that but it resulted to a compilation error in one of our toolchains, can't remember which

guilpier-code commented on src/io/inputs/data-series-csv-importer/DataSeriesRepoImporter.cpp:
```diff
@@ -19,116 +19,142 @@
 }
 
-static TimeSeriesSet matrixToTimeSeriesSet(std::string id, std::vector<std::vector<double>> matrix)
+std::vector<std::vector<double>> readCSV(const std::filesystem::path& filename, char sep)
```
This function could even be **static**

a-zakir committed:
upd

a-zakir committed:
upd

guilpier-code commented on src/io/inputs/data-series-csv-importer/DataSeriesRepoImporter.cpp:
```diff
@@ -19,116 +19,142 @@
-    for (int i = 0; i < nSets; ++i)
+
+    std::vector<std::vector<double>> columns;
+    const char* start = file.data();
```
```c++
const char* start = file.data();
```
**start** is **const** but we change it a bit further. This is strange, but ok.

flomnes replied:
Well that's the difference between `const char*` and `char * const`

guilpier-code commented on src/io/inputs/data-series-csv-importer/DataSeriesRepoImporter.cpp:
```diff
@@ -19,116 +19,142 @@
+            line_len--;
+        }
+
+        auto row = parseNumbersFast(start, start + line_len, sep, columns.size(), fileName);
```
**parseNumbersFast** ==> **parseRowOfNumbers** would be better

a-zakir replied:
row of number does not make sense, what about parseRow ?

guilpier-code commented on src/io/inputs/data-series-csv-importer/DataSeriesRepoImporter.cpp:
```diff
@@ -19,116 +19,142 @@
-        std::vector<double> set;
-        set.reserve(nTimesteps);
-        for (int j = 0; j < nTimesteps; ++j)
+        const char* newline = static_cast<const char*>(memchr(start, '\n', end - start));
```
```c++
const char* newline = ...
```
It seems that **newline** points actually at the end of a line. 
If it's true, why not call it **endLine** ?

a-zakir replied:
done

guilpier-code commented on src/io/inputs/data-series-csv-importer/DataSeriesRepoImporter.cpp:
```diff
@@ -19,116 +19,142 @@
+    {
+        return {}; // empty or inaccessible
+    }
+    const auto& fileName = filename.string();
```
**fileName** : is mainly defined to supply exceptions raised with the current file name.
So instead of using **fileName**  in **readCSV(...)** or passing it further deep, you could try (if possible) to catch the exception thrown, and then rethrow it with the same message enriched with the file name.
To be more specific, we could try something like : 
```c++
try
{
    auto timeSeriesSet = std::make_unique<TimeSeriesSet>(entry.path().stem().string(),
                                                         readCSV(entry, csvSeparator));
}
catch(std::exception& e)
{
    std::string msg = entry.path().stem().string() + " : " + e.what();
    // rethrow the same exception type with the new msg
}
```
If this is not doable easily, then forget it.

a-zakir replied:
I see what you mean. The only reason fileName is supplied to parseNumbersFast and parseOneDouble is to comply with existing tests, which are verifying exception messages.
I think it's not the best use of <code>try/catch</code> to 'just' append the error message.

guilpier-code committed:
Removes temp directories in unit tests (#3100)

in unit tests, some temporary files and directories are never destroyed
after use.

Co-authored-by: guilpier <guillaume.pierre_externe@rte-france.com>

a-zakir committed:
upd

a-zakir committed:
Merge branch 'feature/loadModelerPerf' of https://github.com/AntaresSimulatorTeam/Antares_Simulator into feature/loadModelerPerf

a-zakir committed:
upd

a-zakir committed:
upd

flomnes commented on src/optimisation/linear-problem-data-impl/timeSeriesSet.cpp:
```diff
@@ -21,6 +21,7 @@
 
 #include "antares/optimisation/linear-problem-data-impl/timeSeriesSet.h"
 
+#include <absl/strings/internal/str_format/extension.h>
```
unused

flomnes replied:
```suggestion
```

flomnes commented on src/optimisation/linear-problem-data-impl/timeSeriesSet.cpp:
```diff
@@ -35,6 +36,16 @@ TimeSeriesSet::TimeSeriesSet(std::string name, unsigned int height):
+}
+
 void TimeSeriesSet::add(const std::vector<double>& ts)
 {
```
At line 55 `std::move` will fall back to a copy (ts is a const ref)

a-zakir replied:
this method is not modified by this pr

a-zakir committed:
upd

flomnes commented on src/io/inputs/data-series-csv-importer/DataSeriesRepoImporter.cpp:
```diff
@@ -19,116 +19,149 @@
-    if (!infile.is_open())
+    std::vector<double> row;
+    if (capacity > 0)
     {
-        throw std::invalid_argument("Could not open file " + path.filename().string());
+        row.reserve(capacity);
     }
```
If `capacity=0`, this is a no-op.
```suggestion
    row.reserve(capacity);
```

a-zakir committed:
upd

a-zakir committed:
upd

a-zakir committed:
upd

flomnes commented on src/io/inputs/data-series-csv-importer/DataSeriesRepoImporter.cpp:
```diff
@@ -19,116 +19,158 @@
+    auto sz = std::filesystem::file_size(filename, ec);
+    if (ec || sz == 0)
     {
-        return timeSeriesSet;
+        return {}; // empty or inaccessible
     }
```
IMO we should also throw an exception here, but display a more specific error message

the `return {}` is dangerous, it could lead to invisible errors.

My suggestion 
```cpp
if (sz == 0) return {}; // empty file -> empty matrix
if (ec) throw std::runtime_error("Error reading file " + filename + " (" + ec.message() +")");
```

flomnes commented on src/io/inputs/data-series-csv-importer/DataSeriesRepoImporter.cpp:
```diff
@@ -19,116 +19,158 @@
+    if (rowIndex != 0 && colIndex != columns.size())
+    {
+        throw std::invalid_argument(errorMessagePrefix
+                                    + ": rows have inconsistent number of columns");
```
I suggest choosing a different error message to make debugging easier (see exception at l. 109)

a-zakir committed:
upd
